File: Server/src/main/java/com/github/dr/rwserver/util/io/DefaultSerializers.java
package com.github.dr.rwserver.util.io;

import com.github.dr.rwserver.data.global.Settings;
import com.github.dr.rwserver.net.Administration;
import com.github.dr.rwserver.struct.ObjectMap;
import com.github.dr.rwserver.struct.OrderedMap;
import com.github.dr.rwserver.struct.Seq;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
 * @author Dr
 */
@SuppressWarnings("unchecked")
public class DefaultSerializers {
    public static void register(final Settings settings) {
        settings.setSerializer(String.class, new Settings.TypeSerializer<String>() {
            @Override
            public void write(DataOutput stream, String object) throws IOException {
                stream.writeUTF((object == null) ? "" : object);
            }
            @Override
            public String read(DataInput stream) throws IOException {
                /*  47 */
                return stream.readUTF();
            }
        });
        settings.setSerializer(Seq.class, new Settings.TypeSerializer<Seq>() {
            @Override
            public void write(DataOutput stream, Seq object) throws IOException {
                stream.writeInt(object.size());
                if (object.size() != 0) {
                    Settings.TypeSerializer ser = settings.getSerializer(object.get(0).getClass());
                    if (ser == null) {
                        throw new IllegalArgumentException(object.get(0).getClass() + " does not have a serializer registered!");
                    }
                    stream.writeUTF(object.get(0).getClass().getName());
                    for (Object element : object) {
                        ser.write(stream, element);
                    }
                }
            }
            @Override
            public Seq read(DataInput stream) throws IOException {
                try {
                    int size = stream.readInt();
                    Seq arr = new Seq(size);
                    if (size == 0) {
                        return arr;
                    }
                    String type = stream.readUTF();
                    Settings.TypeSerializer ser = settings.getSerializer(DefaultSerializers.lookup(type));
                    if (ser == null) {
                        throw new IllegalArgumentException(type + " does not have a serializer registered!");
                    }
                    for (int i = 0; i < size; i++) {
                        arr.add(ser.read(stream));
                    }
                    return arr;
                } catch (ClassNotFoundException e) {
                    e.printStackTrace();
                    return null;
                }
            }
        });

        settings.setSerializer(ObjectMap.class, new Settings.TypeSerializer<ObjectMap>() {
            @Override
            public void write(DataOutput stream, ObjectMap map) throws IOException {
                stream.writeInt(map.size);
                if (map.size == 0) {
                    return;
                }
                ObjectMap.Entry entry = map.entries().next();
                Settings.TypeSerializer keySer = settings.getSerializer(entry.key.getClass());
                Settings.TypeSerializer valSer = settings.getSerializer(entry.value.getClass());
                if (keySer == null) {
                    throw new IllegalArgumentException(entry.key.getClass() + " does not have a serializer registered!");
                }
                if (valSer == null) {
                    throw new IllegalArgumentException(entry.value.getClass() + " does not have a serializer registered!");
                }
                stream.writeUTF(entry.key.getClass().getName());
                stream.writeUTF(entry.value.getClass().getName());
                for (ObjectMap.Entries entries = map.entries().iterator(); entries.hasNext(); ) {
                    Object e = entries.next();
                    ObjectMap.Entry en = (ObjectMap.Entry)e;
                    keySer.write(stream, en.key);
                    valSer.write(stream, en.value);
                }
            }
            @Override
            public ObjectMap read(DataInput stream) throws IOException {
                try {
                    int size = stream.readInt();
                    ObjectMap map = new ObjectMap();
                    if (size == 0) {
                        return map;
                    }
                    String keyt = stream.readUTF();
                    String valt = stream.readUTF();
                    Settings.TypeSerializer keySer = settings.getSerializer(DefaultSerializers.lookup(keyt));
                    Settings.TypeSerializer valSer = settings.getSerializer(DefaultSerializers.lookup(valt));
                    if (keySer == null) {
                        throw new IllegalArgumentException(keyt + " does not have a serializer registered!");
                    }
                    if (valSer == null) {
                        throw new IllegalArgumentException(valt + " does not have a serializer registered!");
                    }
                    for (int i = 0; i < size; i++) {
                        Object key = keySer.read(stream);
                        Object val = valSer.read(stream);
                        map.put(key, val);
                    }
                    return map;
                } catch (ClassNotFoundException e) {
                    e.printStackTrace();
                    return null;
                }
            }
        });

        settings.setSerializer(OrderedMap.class, new Settings.TypeSerializer<OrderedMap>() {
            @Override
            public void write(DataOutput stream, OrderedMap map) throws IOException {
                stream.writeInt(map.size);
                if (map.size == 0) {
                    return;
                }
                OrderedMap.Entry entry = map.entries().next();
                Settings.TypeSerializer keySer = settings.getSerializer(entry.key.getClass());
                Settings.TypeSerializer valSer = settings.getSerializer(entry.value.getClass());
                if (keySer == null) {
                    throw new IllegalArgumentException(entry.key.getClass() + " does not have a serializer registered!");
                }
                if (valSer == null) {
                    throw new IllegalArgumentException(entry.value.getClass() + " does not have a serializer registered!");
                }
                stream.writeUTF(entry.key.getClass().getName());
                stream.writeUTF(entry.value.getClass().getName());
                for (OrderedMap.Entries entries = map.entries().iterator(); entries.hasNext(); ) {
                    Object e = entries.next();
                    OrderedMap.Entry en = (OrderedMap.Entry)e;
                    keySer.write(stream, en.key);
                    valSer.write(stream, en.value);
                }
            }
            @Override
            public OrderedMap read(DataInput stream) throws IOException {
                try {
                    int size = stream.readInt();
                    OrderedMap map = new OrderedMap();
                    if (size == 0) {
                        return map;
                    }
                    String keyt = stream.readUTF();
                    String valt = stream.readUTF();
                    Settings.TypeSerializer keySer = settings.getSerializer(DefaultSerializers.lookup(keyt));
                    Settings.TypeSerializer valSer = settings.getSerializer(DefaultSerializers.lookup(valt));
                    if (keySer == null) {
                        throw new IllegalArgumentException(keyt + " does not have a serializer registered!");
                    }
                    if (valSer == null) {
                        throw new IllegalArgumentException(valt + " does not have a serializer registered!");
                    }
                    for (int i = 0; i < size; i++) {
                        Object key = keySer.read(stream);
                        Object val = valSer.read(stream);
                        map.put(key, val);
                    }
                    return map;
                } catch (ClassNotFoundException e) {
                    e.printStackTrace();
                    return null;
                }
            }
        });

        settings.setSerializer(Administration.PlayerInfo.class, new Settings.TypeSerializer<Administration.PlayerInfo>() {
            @Override
            public void write(DataOutput stream, Administration.PlayerInfo object) throws IOException {
                settings.getSerializer(String.class).write(stream, object.uuid);
                stream.writeLong(object.timesKicked);
                stream.writeLong(object.timesJoined);
                stream.writeLong(object.timeMute);
                stream.writeBoolean(object.admin);
            }

            @Override
            public Administration.PlayerInfo read(DataInput stream) throws IOException {
                Administration.PlayerInfo object = new Administration.PlayerInfo((String)settings.getSerializer(String.class).read(stream));
                object.timesKicked = stream.readLong();
                object.timesJoined = stream.readLong();
                object.timeMute = stream.readLong();
                object.admin = stream.readBoolean();
                return object;
            }
        });
    }

    private static Class<?> lookup(String name) throws ClassNotFoundException {
        return Class.forName(name);
   }
}

File: Server-Core/src/main/java/net/rwhps/server/data/plugin/DefaultSerializers.kt
/*
 * Copyright 2020-2023 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package net.rwhps.server.data.plugin

import net.rwhps.server.io.GameInputStream
import net.rwhps.server.io.GameOutputStream
import net.rwhps.server.io.packet.Packet
import net.rwhps.server.net.Administration.PlayerAdminInfo
import net.rwhps.server.net.Administration.PlayerInfo
import net.rwhps.server.struct.ObjectMap
import net.rwhps.server.struct.OrderedMap
import net.rwhps.server.struct.Seq
import net.rwhps.server.struct.SerializerTypeAll.TypeSerializer
import net.rwhps.server.util.algorithms.NetConnectProofOfWork
import java.io.IOException

/**
 * @author RW-HPS/Dr
 */
internal object DefaultSerializers {
    fun register() {
        registrationBasis()
        registerMap()
        registerCustomClass()
    }

    private fun registrationBasis() {
        AbstractPluginData.setSerializer(String::class.java, object : TypeSerializer<String?> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: String?) {
                paramDataOutput.writeString(objectParam ?: "")
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): String {
                return paramDataInput.readString()
            }
        })
        AbstractPluginData.setSerializer(Integer::class.java, object : TypeSerializer<Int> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: Int) {
                paramDataOutput.writeInt(objectParam)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): Int {
                return paramDataInput.readInt()
            }
        })
        AbstractPluginData.setSerializer(java.lang.Float::class.java, object : TypeSerializer<Float> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: Float) {
                paramDataOutput.writeFloat(objectParam)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): Float {
                return paramDataInput.readFloat()
            }
        })
        AbstractPluginData.setSerializer(java.lang.Long::class.java, object : TypeSerializer<Long> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: Long) {
                paramDataOutput.writeLong(objectParam)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): Long {
                return paramDataInput.readLong()
            }
        })
        AbstractPluginData.setSerializer(java.lang.Boolean::class.java, object : TypeSerializer<Boolean> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: Boolean) {
                paramDataOutput.writeBoolean(objectParam)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): Boolean {
                return paramDataInput.readBoolean()
            }
        })
        AbstractPluginData.setSerializer(ByteArray::class.java, object : TypeSerializer<ByteArray> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: ByteArray) {
                paramDataOutput.writeBytesAndLength(objectParam)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): ByteArray {
                return paramDataInput.readStreamBytes()
            }
        })
    }

    private fun registerMap() {
        AbstractPluginData.setSerializer(Seq::class.java, Seq.serializer)

        AbstractPluginData.setSerializer(ObjectMap::class.java, object : TypeSerializer<ObjectMap<*, *>> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: ObjectMap<*, *>) {
                paramDataOutput.writeInt(objectParam.size)
                if (objectParam.size == 0) {
                    return
                }
                val entry = objectParam.entries().next()
                val keySer = AbstractPluginData.getSerializer(entry.key.javaClass) ?: throw getError(entry.key.javaClass.toString())
                val valSer = AbstractPluginData.getSerializer(entry.value.javaClass) ?: throw getError(entry.value.javaClass.toString())
                paramDataOutput.writeString(entry.key.javaClass.name)
                paramDataOutput.writeString(entry.value.javaClass.name)
                for (e in objectParam.entries()) {
                    val en = e as ObjectMap.Entry<*, *>
                    keySer.write(paramDataOutput, en.key)
                    valSer.write(paramDataOutput, en.value)
                }
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): ObjectMap<*, *> {
                val map = ObjectMap<Any?, Any?>()
                try {
                    val size = paramDataInput.readInt()
                    if (size == 0) {
                        return map
                    }
                    val keySerName = paramDataInput.readString()
                    val valSerName = paramDataInput.readString()
                    val keySer = AbstractPluginData.getSerializer(lookup(keySerName)) ?: throw getError(keySerName)
                    val valSer = AbstractPluginData.getSerializer(lookup(valSerName)) ?: throw getError(valSerName)
                    for (i in 0 until size) {
                        val key = keySer.read(paramDataInput)
                        val `val` = valSer.read(paramDataInput)
                        map.put(key, `val`)
                    }
                    return map
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
                return map
            }
        })
        AbstractPluginData.setSerializer(OrderedMap::class.java, object : TypeSerializer<OrderedMap<*, *>> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: OrderedMap<*, *>) {
                paramDataOutput.writeInt(objectParam.size)
                if (objectParam.size == 0) {
                    return
                }
                val entry = objectParam.entries().next()
                val keySer = AbstractPluginData.getSerializer(entry.key.javaClass) ?: throw getError(entry.key.javaClass.toString())
                val valSer = AbstractPluginData.getSerializer(entry.value.javaClass) ?: throw getError(entry.value.javaClass.toString())
                paramDataOutput.writeString(entry.key.javaClass.name)
                paramDataOutput.writeString(entry.value.javaClass.name)
                for (e in objectParam.entries()) {
                    val en = e as ObjectMap.Entry<*, *>
                    keySer.write(paramDataOutput, en.key)
                    valSer.write(paramDataOutput, en.value)
                }
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): OrderedMap<*, *> {
                val map = OrderedMap<Any?, Any?>()
                try {
                    val size = paramDataInput.readInt()
                    if (size == 0) {
                        return map
                    }
                    val keySerName = paramDataInput.readString()
                    val valSerName = paramDataInput.readString()
                    val keySer = AbstractPluginData.getSerializer(lookup(keySerName)) ?: throw getError(keySerName)
                    val valSer = AbstractPluginData.getSerializer(lookup(valSerName)) ?: throw getError(valSerName)
                    for (i in 0 until size) {
                        val key = keySer.read(paramDataInput)
                        val `val` = valSer.read(paramDataInput)
                        map.put(key, `val`)
                    }
                    return map
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
                return map
            }
        })
    }

    private fun registerCustomClass() {
        AbstractPluginData.setSerializer(PlayerInfo::class.java, object : TypeSerializer<PlayerInfo> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: PlayerInfo) {
                AbstractPluginData.getSerializer(String::class.java)!!.write(paramDataOutput, objectParam.uuid)
                paramDataOutput.writeLong(objectParam.timesKicked)
                paramDataOutput.writeLong(objectParam.timesJoined)
                paramDataOutput.writeLong(objectParam.timeMute)
                paramDataOutput.writeBoolean(objectParam.admin)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): PlayerInfo {
                val objectParam = PlayerInfo(AbstractPluginData.getSerializer(String::class.java)!!.read(paramDataInput) as String)
                objectParam.timesKicked = paramDataInput.readLong()
                objectParam.timesJoined = paramDataInput.readLong()
                objectParam.timeMute = paramDataInput.readLong()
                objectParam.admin = paramDataInput.readBoolean()
                return objectParam
            }
        })

        AbstractPluginData.setSerializer(PlayerAdminInfo::class.java, object : TypeSerializer<PlayerAdminInfo> {
            @Throws(IOException::class)
            override fun write(paramDataOutput: GameOutputStream, objectParam: PlayerAdminInfo) {
                paramDataOutput.writeString(objectParam.uuid)
                paramDataOutput.writeBoolean(objectParam.admin)
                paramDataOutput.writeBoolean(objectParam.superAdmin)
            }

            @Throws(IOException::class)
            override fun read(paramDataInput: GameInputStream): PlayerAdminInfo {
                return PlayerAdminInfo(paramDataInput.readString(), paramDataInput.readBoolean(), paramDataInput.readBoolean())
            }
        })

        AbstractPluginData.setSerializer(Packet::class.java, Packet.serializer)
        AbstractPluginData.setSerializer(NetConnectProofOfWork::class.java, NetConnectProofOfWork.serializer)
    }

    @Throws(ClassNotFoundException::class)
    fun lookup(name: String): Class<*> {
        return Class.forName(name)
    }

    fun getError(className: String): IllegalArgumentException {
        return IllegalArgumentException("$className does not have a serializer registered!")
    }
}